src/mds/MDCache.cc

```diff
@@ -12969,6 +12972,167 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
   mds->server->respond_to_request(mdr, r);
 }
 
+class C_MDC_DataUninlinedSubmitted : public MDCacheLogContext {
+  MDRequestRef mdr;
+
+  public:
+  C_MDC_DataUninlinedSubmitted(MDRequestRef r) :
+    MDCacheLogContext(static_cast<CInode*>(r->internal_op_private)->mdcache), mdr(r) {}
+
+  void finish(int r) {
+    auto mds = get_mds(); // to keep dout happy
+
+    dout(20) << "(uninline_data) log submission " << (r ? "failed" : "succeeded") << dendl;
+
+    mdr->apply();
+    mdr->internal_op_finish->complete(0);
+    mdr->internal_op_finish = new C_MDSInternalNoop;
+    mds->server->respond_to_request(mdr, r);
+  }
+};
+
+struct C_IO_DataUninlined : public Context {
+  MDRequestRef mdr;
+
+  public:
+  C_IO_DataUninlined(MDRequestRef r) : mdr(r) {}
+
+  void finish(int r) override {
+    auto in = static_cast<CInode*>(mdr->internal_op_private);
+    auto mds = in->mdcache->mds;
+
+    // return faster if operation has failed (non-zero) status
+    if (r) {
+      dout(20) << "(uninline_data) mutation failed" << dendl;
+      mds->server->respond_to_request(mdr, r);
+      return;
+    }
+
+    dout(20) << "(uninline_data) mutation succeeded for " << *in << dendl;
+
+    // journal the inode changes
+    MDLog *mdlog = mds->mdlog;
+
+    dout(20) << "(uninline_data) writing to journal for " << *in << dendl;
+
+    {
+      EUpdate *le = new EUpdate(mdlog, "uninline");
+      // entity_name_t entity = entity_name_t::MDS(mds->get_nodeid());
+      // metareqid_t reqid(entity, mds->get_next_internal_reqid());
+
+      // le->metablob.add_client_req(reqid);
+      mdr->ls = mdlog->get_current_segment();
+
+      std::scoped_lock mds_lock(mds->mds_lock);
+      
+      mdlog->start_entry(le);
+      
+      auto pi = in->project_inode(mdr);
+      pi.inode->version = in->pre_dirty();
+      pi.inode->inline_data.free_data();
+      pi.inode->inline_data.version = CEPH_INLINE_NONE;
+      pi.inode->ctime = mdr->get_op_stamp();
+      if (mdr->get_op_stamp() > pi.inode->rstat.rctime)
+	pi.inode->rstat.rctime = mdr->get_op_stamp();
+      pi.inode->change_attr++;
+
+      in->mdcache->predirty_journal_parents(mdr, &le->metablob, in, nullptr,
+					    PREDIRTY_PRIMARY);
+      in->mdcache->journal_dirty_inode(mdr.get(), &le->metablob, in);
+
+      mdr->committing = true;
+
+      string event_str("submit entry: ");
+      event_str += __func__;
+      mdr->mark_event(event_str);
+
+      auto fin = new C_MDC_DataUninlinedSubmitted(mdr);
+      mdlog->submit_entry(le, fin);
```

Comment:
IMO you should journal the request in `uninline_data_work()` instead just before you do anything as early as possible.

Reply:
any other sequence of journaling (which requires mds_lock) causes crashes if not done atomically under the mds_lock

Reply:
When the `uninline_data_work()` is called the `mds_lock` has already been hold. For this you could refer to `Server::handle_client_setattr()` just follow what the inode truncate does.

Reply:
`uninline_data_work()` is an asynchronous activity. At the end of the function, a continuation context is passed to the objecter trying to mutate the file. When the objecter finishes its job, the continuation is called to notify work completion and then the event is recorded in the journal.

I tried to start logging the event in `uninline_data_work()` as you suggested, but the mds crashed with a traceback when trying to start another entry ... this is because although `uninline_data_work()` had acquired the `mds_lock`, further work was delegated to an asynchronous objecter activity which caused the `mds_lock` to be released since the `scoped_lock` lost its scope. Do you recommend locking the `mds_lock` in `uninline_data_work()` and unlock in `C_IO_DataUninlined::finish()` ? I ran a test with these changes, but neither the MDS crashed nor were the files uninlined.

The inode truncate workflow in `Server::handle_client_setattr()` is a bit different since all the data IO activity has been taken care of by the client by interacting directly with the OSDs and has only delegated updating attr metadata to the MDS. In case of `uninline_data_work()` the file data IO needs to be taken care of by the MDS as well followed by metadata update and logging to the journal.

Reply:
wouldn't writing data blocks later be a problem is there's insufficient space to write out the data blocks after journaling the changes apriori ?

on the other hand, if the journal writing fails, then clients can continue to read from the inode as usual even if the blocks have been written to the 0th file object in the data pool

Reply:
> wouldn't writing data blocks later be a problem if there's insufficient space to write out the data blocks after journaling the changes apriori ?

You should check it before journaling.

> 
> on the other hand, if the journal writing fails, then clients can continue to read from the inode as usual even if the blocks have been written to the 0th file object in the data pool

We shouldn't be care about the journal failing case, or the other normal cases will have the same issue, and how do they handle it ?

Thinking more about it, this is a little special. Because it not only will handle the metadatas but also the datas of a file. If we journal it and then set the `version` to `CEPH_INLINE_NONE` and then return success, then when a client want to read it and if the journals are still not expired yet. The read may fail or just return 0, because only after the journals expired will the `0th` object created and have the inline data.

So this should be okay, we shouldn't journal it.